crates/pathfinder/src/ethereum/api.rs

```diff
@@ -0,0 +1,246 @@
+//! Wrapper for the parts of the [`Web3::eth()`](https://docs.rs/web3/latest/web3/api/struct.Eth.html) API that [the ethereum module](super) uses.
+use crate::retry::Retry;
+
+use std::future::Future;
+use std::num::NonZeroU64;
+use std::time::Duration;
+
+use futures::TryFutureExt;
+use tracing::{debug, error, info};
+use web3::{
+    types::{Block, BlockId, Filter, Log, Transaction, TransactionId, H256, U256},
+    Error, Transport, Web3,
+};
+
+/// Error returned by [`Web3EthImpl::logs`].
+#[derive(Debug, thiserror::Error)]
+pub enum LogsError {
+    /// Query exceeded limits (time or result length).
+    #[error("Query limit exceeded.")]
+    QueryLimit,
+    /// One of the blocks specified in the filter is unknown. Currently only
+    /// known to occur for Alchemy endpoints.
+    #[error("Unknown block.")]
+    UnknownBlock,
+    #[error(transparent)]
+    Other(#[from] web3::Error),
+}
+
+#[async_trait::async_trait]
+/// Contains only those functions from [`Web3::eth()`](https://docs.rs/web3/latest/web3/api/struct.Eth.html)
+/// that [the ethereum module](super) uses.
+pub trait Web3EthApi {
+    async fn block(&self, block: BlockId) -> web3::Result<Option<Block<H256>>>;
+    async fn block_number(&self) -> web3::Result<u64>;
+    async fn chain_id(&self) -> web3::Result<U256>;
+    async fn logs(&self, filter: Filter) -> std::result::Result<Vec<Log>, LogsError>;
+    async fn transaction(&self, id: TransactionId) -> web3::Result<Option<Transaction>>;
+}
+
+/// An implementation of [`Web3EthApi`] which uses [`Web3::eth()`](https://docs.rs/web3/latest/web3/api/struct.Eth.html)
+/// wrapped in an [exponential backoff retry utility](Retry).
+///
+/// Initial backoff time is 30 seconds and saturates at 1 hour:
+///
+/// `backoff [secs] = min((2 ^ N) * 15, 3600) [secs]`
+///
+/// where `N` is the consecutive retry iteration number `{1, 2, ...}`.
+#[derive(Clone, Debug)]
+pub struct Web3EthImpl<T>(pub Web3<T>)
```

Comment:
I don't think we have any need to be generic over the transport internally as we can always change as long as ... Are both of the web3 transports Send and Sync, but they had just left the requirement out of Transport?

Reply:
- All [transports](https://docs.rs/web3/latest/web3/transports/index.html) are `Send` and `Sync`.
- Thanks for noticing the `pub`, I was thinking about it and forgot later on. I can add a `fn new` :+1: . I'm using `.0` notation in some tests though but I can add a `#[cfg(test)] impl Deref` for that.
- I made the transport generic anticipating `ws` could be used as well, but this can always be changed from `Http`, so I'll just make it `Http`.